### ISSUE TITLE
[Feature] Add checkpoint_best_only in train_grad

### ISSUE THREAD

**Issue by dominikandreasseitz** [pasqal-io/qadence#214]
Qadence offers a `train_grad` ([link](https://github.com/pasqal-io/qadence/blob/main/qadence/ml_tools/train_grad.py)) method which abstracts some QNN training-related tasks away from the user.
It expects a `TrainConfig` ([link](https://github.com/pasqal-io/qadence/blob/main/qadence/ml_tools/config.py#L39)) which can be used to pass number of training epochs and various other training-related hyperparameters.

Task:
1. Implement logic in `train_grad` which only saves a checkpoint if the loss has improved by some user-provided `epsilon`. This can for example be done via maintaining a `current_best_validation_loss` and only save a checkpoint if `current_loss <= current_best_validation_loss - epsilon ` 
2. Add a test which makes sure your solution works
3. Add an example in our [tutorials](https://github.com/pasqal-io/qadence/blob/main/docs/tutorials/advanced_tutorials/custom-models.md) showcasing how to use it

**Comment by madagra** [comment]
@dominikandreasseitz how do you define the "best" exactly? Given the creation date of the checkpoint, for example?

**Comment by dominikandreasseitz** [comment]
@madagra i added a description

**Comment by Roland-djee** [comment]
@dominikandreasseitz Can this be closed ?

**Comment by dominikandreasseitz** [comment]
@Roland-djee no, needs to be added

**Comment by Yash-10** [comment]
Hello @dominikandreasseitz , I have been trying to implement this, but I have a few questions. Could you help?

1. Is it intended to add `checkpoint_best_only` based on the train or the validation loss? I believe it's the latter since that would make more sense. If so, perhaps a new argument to the `train()` function can be added that accepts a validation data loader.
2. What is the use case of `DictDataLoader` as opposed to the (regular) `DataLoader`? (I believe this is not explained in the docs in detail, but I may have missed something).
    - For example, is it assumed that for a `DictDataLoader`, the first data loader should be used for training and the other for validation? I believe not because even if the data loader being passed is of type `DictDataLoader`, it is passed entirely to `optimize_step`. But if it's yes, then perhaps the validation can be run whenever the data loader being passed is of type `DictDataLoader`.

**Comment by dominikandreasseitz** [comment]
> Hello @dominikandreasseitz , I have been trying to implement this, but I have a few questions. Could you help?
> 
> 1. Is it intended to add `checkpoint_best_only` based on the train or the validation loss? I believe it's the latter since that would make more sense. If so, perhaps a new argument to the `train()` function can be added that accepts a validation data loader.
> 2. What is the use case of `DictDataLoader` as opposed to the (regular) `DataLoader`? (I believe this is not explained in the docs in detail, but I may have missed something).
>    
>    * For example, is it assumed that for a `DictDataLoader`, the first data loader should be used for training and the other for validation? I believe not because even if the data loader being passed is of type `DictDataLoader`, it is passed entirely to `optimize_step`. But if it's yes, then perhaps the validation can be run whenever the data loader being passed is of type `DictDataLoader`.

hi @Yash-10, thanks for participating! i assigned the issue to you. 
Regarding your quesitons: Ideally we want the validation loss as a criterion to save checkpoints, you're right.
Given the current interface, i think its best to do the following:

- Modify the `train` method with a condition like:
```
                if iteration % config.val_every == 0:
                    val_loss = loss_fn(model,...)
                    metrics['val_loss'] = val_loss
                    write_tensorboard(writer, None, metrics, iteration)
```
in that sense, you can set the frequency of the validation calculation using `val_every` and just take the validation data from the `dataloader` object.

what do you think?

**Comment by Yash-10** [comment]
Thank you for your reply! Indeed, that helps me gain more clarity. We can maybe continue in #435 to discuss this further in more context.

**Comment by Yash-10** [comment]
@dominikandreasseitz thank you so much for merging my PR recently. Do you think this issue can be closed?
